src/record/kv.rs

```diff
@@ -0,0 +1,290 @@
+        if let Some((key, value)) = first {
+            if brackets {
+                dest.write_str(" { ")?;
+            }
+
+            // Reduce branch prediction misses for performance
+            // So we manually process the first KV pair
+            dest.write_str(key.as_str())?;
+            dest.write_str("=")?;
+            write!(dest, "{value}")?;
+
+            for (key, value) in iter {
+                dest.write_str(" ")?;
+                dest.write_str(key.as_str())?;
+                dest.write_str("=")?;
+                write!(dest, "{value}")?;
+            }
+
+            if brackets {
+                dest.write_str(" }")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<'a> IntoIterator for KeyValues<'a> {
+    type Item = Pair<'a>;
+    type IntoIter = KeyValuesIter<'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+/// Represents an iterator over key-value pairs.
+pub struct KeyValuesIter<'a>(KeyValuesIterInner<'a>);
+
+impl<'a> Iterator for KeyValuesIter<'a> {
+    type Item = Pair<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match &mut self.0 {
+            // The 2 clones should be cheap
+            KeyValuesIterInner::Borrowed(iter) => iter.next().map(|(k, v)| (k.clone(), v.clone())),
+            KeyValuesIterInner::Owned(iter) => iter.next().map(|(k, v)| (k.as_ref(), v.by_ref())),
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        match &self.0 {
+            KeyValuesIterInner::Borrowed(iter) => iter.size_hint(),
+            KeyValuesIterInner::Owned(iter) => iter.size_hint(),
+        }
+    }
+}
+
+pub(crate) type Pair<'a> = (Key<'a>, Value<'a>);
+
+pub(crate) struct LogCrateConverter<'a>(Vec<(log::kv::Key<'a>, ValueOwned)>);
+
+impl<'a> LogCrateConverter<'a> {
+    pub(crate) fn new(capacity: usize) -> Self {
+        Self(Vec::with_capacity(capacity))
+    }
+
+    pub(crate) fn finalize(self) -> Vec<(log::kv::Key<'a>, ValueOwned)> {
+        self.0
+    }
+}
+
+impl<'a> log::kv::VisitSource<'a> for LogCrateConverter<'a> {
+    fn visit_pair(
+        &mut self,
+        key: log::kv::Key<'a>,
+        value: log::kv::Value<'a>,
+    ) -> Result<(), log::kv::Error> {
+        struct Visitor(Option<ValueOwned>);
+
+        macro_rules! visit_fn {
+            ( $($fn:ident: $ty:ty => $from:ident),+$(,)? ) => {
+                $(fn $fn(&mut self, value: $ty) -> Result<(), log::kv::Error> {
+                    self.0 = Some(Value::$from(value).to_owned());
+                    Ok(())
+                })+
+            };
+        }
+
+        impl log::kv::VisitValue<'_> for Visitor {
+            fn visit_any(&mut self, value: log::kv::Value) -> Result<(), log::kv::Error> {
+                // Since we have no way to extract the underlying `&dyn Display`, we have to
+                // `to_owned()` here
+                self.0 = Some(Value::from_display(&value).to_owned());
```

Comment:
This is a place where you could use `serde` or `sval` to visit the `log` value and serialize it into your value type of choice.

I think we could add some methods to `value_bag::OwnedValueBag` to let you buffer in a value directly.

Reply:
Ah, I see that doesn't win us any more than `ValueBag::from_dyn_sval1(value).to_owned()`, which will also buffer a structured `log::kv::Value`.

Reply:
Yeah. I've integrated with the `Str` abstraction and notice that log's KV can hardly be adapted with zero overhead, especially because the value is highly encapsulated behind the trait and visitor ...

Reply:
Let me try with `kv::Value::from_sval2(v)`, while it adds new dependency to sval, but log's kv already more or less depends on this.